---
title: 报错处理
icon: markdown
date: 2024-04-08
---

## 开启立即监听，响应式数据的引用问题

### 问题描述

需求是监听 tab，发送不同的请求。并且需要在初始阶段就请求一次（想省去 onMounted），所以需要开启立即监听。

但是开启立即监听后，出现了 `Cannot access 'searchForm' before initialization` 的报错。

这是因为在 watch hook 函数中引用了 searchForm 这个响应式数据，而该数据定义在 watch 后面，如果不开立即监听就没关系，一旦开启立即监听后，就会报错。

```ts
const tab = ref("ON_SHELF")

watch(tab, (tab: string) => {
  switch (tabName) {
    case "ON_SHELF":
      getGoodsInfo(1)
      break
    case "OFF_SHELF":
      getGoodsInfo(2)
      break
    case "REJECTED":
      getGoodsInfo(3)
      break
    case "IN_REVIEW":
      getGoodsInfo(4)
      break
  }
}, {
  immediate: true
})

const searchForm = reactive({
  /* ... */
})

async function getGoodsInfo(status: number) {
  const response = await apiGoodsInfo({
    ...searchForm, // Cannot access 'searchForm' before initialization
    status
  })
  
  if (response.data.code === RESPONSE_CODE.SUCCESS) {
    goodsInfo.value = response.data.data
  }
}
```

### 测试

这里将代码简化，做一个测试。响应式数据 a 定义在 watch 的下方，当在 watch hook 中引用该数据并开启立即监听时，出现了 `Cannot access 'a' before initialization` 的报错。

```ts
const forWatch = ref("")

watch(forWatch, () => {
  console.log(a.value) // Cannot access 'a' before initialization
}, {
  immediate: true
})

const a = ref(1)
```

而将响应式数据 a 定义在 watch 的上方时，就不会出现这个报错，并打印了 a 的值。

```ts
const forWatch = ref("")

const a = ref(1)

watch(forWatch, () => {
  console.log(a.value) // 1
}, {
  immediate: true
})
```

如果 a 是一个普通数据，而不是响应式数据。测试发现，即使将 a 定义在 watch 的下方，也没有出现报错，并且也打印了 a 的值。

```ts
const forWatch = ref("")

watch(forWatch, () => {
  console.log(a) // 1
}, {
  immediate: true
})

const a = 1
```

### 结论

使用 watch 的立即监听时，不能在 watch hook 中引用定义在其之后的响应式数据（ref，reactive），

但是可以引用定义在该 watch 之后的普通数据。

如果 watch 不开启立即监听，那么就可以引用定义在该 watch 之后的响应式数据和普通数据。
